src/backend/distributed/commands/alter_table.c

```diff
@@ -0,0 +1,811 @@
+		.shardCount = 0,
+		.colocateWith = NULL,
+		.accessMethod = accessMethod,
+		.cascadeToColocated = false
+	};
+	ConvertTable(config);
+}
+
+
+/*
+ * ConvertTable is used for converting a table into a new table with different properties.
+ * The conversion is done by creating a new table, moving everything to the new table and
+ * dropping the old one. So the oid of the table is not preserved.
+ *
+ * The new table will have the same name, columns and rows. It will also have partitions,
+ * views, sequences of the old table. Finally it will have everything created by
+ * GetPostLoadTableCreationCommands function, which include indexes. These will be
+ * re-created during conversion, so their oids are not preserved either (except for
+ * sequences). However, their names are preserved.
+ *
+ * The dropping of old table is done with CASCADE. Anything not mentioned here will
+ * be dropped.
+ */
+void
+ConvertTable(TableConversionConfiguration config)
+{
+	List *colocatedTableList = NIL;
+	if (config.cascadeToColocated)
+	{
+		colocatedTableList = ColocatedTableList(config.relationId);
+	}
+
+	bool shardCountIsNull = false;
+	if (config.shardCount == 0 && IsCitusTableType(config.relationId, DISTRIBUTED_TABLE))
+	{
+		CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(config.relationId);
+		config.shardCount = cacheEntry -> shardIntervalArrayLength;
+		shardCountIsNull = true;
+	}
+	List *preLoadCommands = GetPreLoadTableCreationCommands(config.relationId, true, config.accessMethod);
+	List *postLoadCommands = GetPostLoadTableCreationCommands(config.relationId);
+	List *justBeforeDropCommands = NIL;
+
+	postLoadCommands = list_concat(postLoadCommands,
+								   GetViewCreationCommandsOfTable(config.relationId));
+
+	char *relationName = get_rel_name(config.relationId);
+	Oid schemaId = get_rel_namespace(config.relationId);
+	char *schemaName = get_namespace_name(schemaId);
+
+	int spiResult = SPI_connect();
+	if (spiResult != SPI_OK_CONNECT)
+	{
+		ereport(ERROR, (errmsg("could not connect to SPI manager")));
+	}
+
+	bool isPartitionTable = false;
+	char *attachToParentCommand = NULL;
+	if (PartitionTable(config.relationId))
+	{
+		isPartitionTable = true;
+		char *detachFromParentCommand = GenerateDetachPartitionCommand(config.relationId);
+		attachToParentCommand = GenerateAlterTableAttachPartitionCommand(config.relationId);
+
+
+		justBeforeDropCommands = lappend(justBeforeDropCommands, detachFromParentCommand);
+	}
+
+	if (PartitionedTable(config.relationId))
+	{
+		ereport(NOTICE, (errmsg("undistributing the partitions of %s",
+								quote_qualified_identifier(schemaName, relationName))));
+		List *partitionList = PartitionList(config.relationId);
+		Oid partitionRelationId = InvalidOid;
+		foreach_oid(partitionRelationId, partitionList)
+		{
+			char *detachPartitionCommand = GenerateDetachPartitionCommand(
+				partitionRelationId);
+			char *attachPartitionCommand = GenerateAlterTableAttachPartitionCommand(
```

Comment:
we seem to have detach/attach logic for both partitions and partitioned tables. Do we need both?

Reply:
One is for when a table is parent partitioned table (`undistribute_table('partitioned_table')`) and the other is when the table is child partition table (`alter_table_set_access_method('partition_table', 'columnar')`). So they are for different cases. I guess I could try to find a way to combine them.

Reply:
Makes sense, one thing I missed is that we already detached the partition before we get to `if (PartitionTable(con->relationId))` in the nested call. Would be good to add a comment on this.